File: 周赛/第 472 场周赛/q2.py
from typing import List, Tuple, Dict, Set, Optional
from collections import defaultdict, Counter, deque
from itertools import permutations, combinations, accumulate
from datetime import datetime, date, time, timedelta
from time import time as timestamp, sleep
from functools import cmp_to_key, lru_cache, reduce
from math import gcd, sqrt, log, ceil, floor, inf
from bisect import bisect_left, bisect_right
from heapq import heappush, heappop, heapify, nsmallest, nlargest
from sortedcontainers import SortedList
from sys import setrecursionlimit
setrecursionlimit(5 * 10 ** 5 + 1)
def fmax(a, b): return a if a > b else b
def fmin(a, b): return a if a < b else b
def lcm(a, b): return a * b // gcd(a, b)
MOD = 10 ** 9 + 7

class Solution:
    def longestBalanced(self, nums: List[int]) -> int:
        n = len(nums)
        ans = 0
        
        for i in range(n):
            cnt0 = set()
            cnt1 = set()
            for j in range(i, n):
                num = nums[j]
                if num % 2 == 0:
                    cnt0.add(num)
                else:
                    cnt1.add(num)
                
                if len(cnt0) == len(cnt1):
                    ans = fmax(ans, j - i + 1)
    
        return ans
    
if __name__ == "__main__":
    s = Solution()
    print(s.longestBalanced(nums = [3,2,2,5,4]))
    print(s.longestBalanced(nums = [10,6,10,7]))

File: 模拟面试/Ⅲ/Q2.cpp
#include <bits/stdc++.h>
using namespace std;
const int MOD = 1e9 + 7;

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int n = nums.size(), ans = 0;
        unordered_map<int, int> dic;
        vector<int> pre(n + 1, 0);
        for (int i = 0; i < n; i++) {
            pre[i + 1] = pre[i] + nums[i];
        }
        for (int s : pre) {
            ans += dic[s - k];
            dic[s] += 1;
        }
        return ans;
    }
};